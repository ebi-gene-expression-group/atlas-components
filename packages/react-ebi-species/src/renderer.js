"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
