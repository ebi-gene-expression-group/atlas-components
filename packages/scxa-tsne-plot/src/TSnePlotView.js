import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'

import ClusterTSnePlot from './ClusterTSnePlot'
import GeneExpressionTSnePlot from './GeneExpressionTSnePlot'

const fetchResponseJson = async (base, endpoint) => {
  const response = await fetch(URI(endpoint, base).toString())
  const responseJson = await response.json()
  return responseJson
}

class TSnePlotView extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      geneExpressionData: {
        max: null,
        min: null,
        series: [],
        unit: ``
      },
      cellClustersData: {
        series: []
      },
      geneExpressionErrorMessage: null,
      cellClustersErrorMessage: null,
      loadingCellClusters: false,
      loadingGeneExpression: false
    }
  }

  _fetchAndSetStateCellClusters({atlasUrl, experimentAccession, selectedColourBy, selectedColourByCategory, selectedPerplexity}) {
    this.setState({
      loadingCellClusters: true
    }, () => {
      let endpoint
      if(selectedColourByCategory === `clusters`) {
        endpoint = `json/experiments/${experimentAccession}/tsneplot/${selectedPerplexity}/clusters/k/${selectedColourBy}`
      }
      else if(selectedColourByCategory === `metadata`) {
        endpoint = `json/experiments/${experimentAccession}/tsneplot/${selectedPerplexity}/metadata/${selectedColourBy}`
      }
      fetchResponseJson(atlasUrl, endpoint)
        .then((responseJson) => {
          this.setState({
            cellClustersData: responseJson,
            cellClustersErrorMessage: null,
            loadingCellClusters: false
          })
        })
        .catch((reason) => {
          this.setState({
            cellClustersErrorMessage: `${reason.name}: ${reason.message}`,
            loadingCellClusters: false
          })
        })
    })
  }

  _fetchAndSetStateGeneId({atlasUrl, experimentAccession, selectedPerplexity, geneId}) {
    this.setState({
      loadingGeneExpression: true
    }, () => {
      fetchResponseJson(atlasUrl, `json/experiments/${experimentAccession}/tsneplot/${selectedPerplexity}/expression/${geneId}`)
        .then((responseJson) => {
          this.setState({
            geneExpressionData: responseJson,
            geneExpressionErrorMessage: null,
            loadingGeneExpression: false,
          })
        })
        .catch((reason) => {
          this.setState({
            geneExpressionErrorMessage: `${reason.name}: ${reason.message}`,
            loadingGeneExpression: false
          })
        })
    })
  }

  componentDidUpdate(previousProps) {
    if (previousProps.selectedPerplexity !== this.props.selectedPerplexity || previousProps.experimentAccession !== this.props.experimentAccession) {
      this._fetchAndSetStateCellClusters(this.props)
      this._fetchAndSetStateGeneId(this.props)
    } else if (previousProps.selectedColourByCategory !== this.props.selectedColourBy && previousProps.selectedColourBy !== this.props.selectedColourBy) {
      this._fetchAndSetStateCellClusters(this.props)
    } else if (previousProps.geneId !== this.props.geneId) {
      this._fetchAndSetStateGeneId(this.props)
    }
  }

  componentDidMount() {
    this._fetchAndSetStateCellClusters(this.props)
    this._fetchAndSetStateGeneId(this.props)
  }

  render() {
    const {height, atlasUrl, resourcesUrl, suggesterEndpoint} = this.props
    const {wrapperClassName, clusterPlotClassName, expressionPlotClassName} = this.props
    const {geneId, speciesName, highlightClusters} = this.props
    const {ks, perplexities, selectedPerplexity, metadata, selectedColourBy} = this.props
    const {onChangePerplexity, onSelectGeneId, onChangeColourBy} = this.props
    const {loadingGeneExpression, geneExpressionData, geneExpressionErrorMessage} = this.state
    const {loadingCellClusters, cellClustersData, cellClustersErrorMessage} = this.state

    const getTooltipContent = (cellId) => {
      return fetchResponseJson(atlasUrl, `json/experiment/${this.props.experimentAccession}/cell/${cellId}/metadata`)
    }

    return (
      <div className={wrapperClassName}>
        <div className={clusterPlotClassName}>
          <ClusterTSnePlot
            height={height}
            plotData={cellClustersData}
            perplexities={perplexities}
            selectedPerplexity={selectedPerplexity}
            onChangePerplexity={onChangePerplexity}
            ks={ks}
            metadata={metadata}
            onChangeColourBy={onChangeColourBy}
            selectedColourBy={selectedColourBy}
            highlightClusters={highlightClusters}
            loading={loadingCellClusters}
            resourcesUrl={resourcesUrl}
            errorMessage={cellClustersErrorMessage}
            tooltipContent={getTooltipContent}
          />
        </div>

        <div className={expressionPlotClassName}>
          <GeneExpressionTSnePlot
            height={height}
            plotData={geneExpressionData}
            atlasUrl={atlasUrl}
            suggesterEndpoint={suggesterEndpoint}
            onSelectGeneId={onSelectGeneId}
            geneId={geneId}
            speciesName={speciesName}
            highlightClusters={[]}
            loading={loadingGeneExpression}
            resourcesUrl={resourcesUrl}
            errorMessage={geneExpressionErrorMessage}
          />
        </div>
      </div>
    )
  }

  componentDidCatch(error) {
    this.setState({
      errorMessage: `${error}`
    })
  }
}

TSnePlotView.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,
  clusterPlotClassName: PropTypes.string,
  expressionPlotClassName: PropTypes.string,
  suggesterEndpoint: PropTypes.string.isRequired,
  experimentAccession: PropTypes.string.isRequired,
  ks: PropTypes.arrayOf(PropTypes.number).isRequired,
  perplexities: PropTypes.arrayOf(PropTypes.number).isRequired,
  selectedPerplexity: PropTypes.number.isRequired,

  metadata: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string
  })),
  selectedColourBy: PropTypes.string,
  selectedColourByCategory: PropTypes.string,
  onChangeColourBy: PropTypes.func,

  highlightClusters: PropTypes.arrayOf(PropTypes.number),
  geneId: PropTypes.string.isRequired,
  speciesName: PropTypes.string.isRequired,
  height: PropTypes.number,
  resourcesUrl: PropTypes.string,
  onSelectGeneId: PropTypes.func,
  onChangePerplexity: PropTypes.func
}

TSnePlotView.defaultProps = {
  highlightClusters: [],
  wrapperClassName: `row`,
  clusterPlotClassName: `small-12 medium-6 columns`,
  expressionPlotClassName: `small-12 medium-6 columns`,
  geneId: ``,
  speciesName: ``,
  height: 800,
  onSelectGeneId: () => {},
  onChangeColourBy: () => {},
  onPerplexityChange: () => {}
}

export default TSnePlotView
