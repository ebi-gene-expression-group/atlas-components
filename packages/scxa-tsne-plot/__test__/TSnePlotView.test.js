import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import '@babel/polyfill'
import fetchMock from 'fetch-mock'

import TSnePlotView from '../src/TSnePlotView'

describe(`TSnePlotView`, () => {

  beforeEach(() => {
    fetchMock.restore()
  })

  const testExperiment = {
    accession: `E-MTAB-5061`,
    accessKey: `f60a21b8-990a-49d9-95aa-623c10865faa`,
    species: `Homo sapiens`,
    ks: [20],
    metadata: [
      {
        value: `inferred cell type - authors labels`,
        label: `inferred cell type - authors labels`
      },
      {
        value: `inferred cell type - ontology labels`,
        label: `inferred cell type - ontology labels`
      }
    ],
    plotTypesAndOptions: {
      "tsne": [{ "perplexity": 40 }, { "perplexity": 25 }, { "perplexity": 45 },{ "perplexity": 1 },{ "perplexity": 30 },
        {"perplexity": 10 },{ "perplexity": 15 },{ "perplexity": 50 },{ "perplexity": 35 },{ "perplexity": 20 },{ "perplexity": 5 }],
      "umap": [{"n_neighbors": 5},{"n_neighbors": 100},{"n_neighbors": 50},{"n_neighbors": 10},{"n_neighbors": 30},{"n_neighbors": 15},{"n_neighbors": 3}]
    }
  }

  const { accession, accessKey, ks, metadata, species, plotTypesAndOptions } = testExperiment

  const plotTypeDropdown =  [
    {
      plotType: `UMAP`,
      plotOptions: plotTypesAndOptions.umap
    },
    {
      plotType: `tSNE`,
      plotOptions: plotTypesAndOptions.tsne
    }
  ]

  test(`matches snapshot`, () => {
    const tree = renderer
      .create(<TSnePlotView
        atlasUrl={``}
        metadata={metadata}
        accessKey={accessKey}
        suggesterEndpoint={``}
        ks={ks}
        geneId={``}
        selectedPlotType={`umap`}
        species={species}
        experimentAccession={accession}
        selectedPlotOption={5}
        plotTypeDropdown={plotTypeDropdown}
        selectedPlotOptionLabel={``}
      />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test(`updates its state with the JSON payload after a successful request`, async () => {
    const payload = {
      series: [`foo`, `bar`]
    }

    fetchMock.get(`foo/json/cell-plots/` + accession + `/clusters/k/20?plotMethod=umap&accessKey=` + accessKey + `&` , JSON.stringify(payload))

    const wrapper = shallow(
      <TSnePlotView
        atlasUrl={`foo/`}
        metadata={metadata}
        accessKey={accessKey}
        suggesterEndpoint={`json/suggestions`}
        ks={ks}
        selectedColourBy={`20`}
        selectedColourByCategory={`clusters`}
        selectedPlotType={`umap`}
        species={species}
        experimentAccession={accession}
        selectedPlotOption={5}
        plotTypeDropdown={plotTypeDropdown}
        selectedPlotOptionLabel={``}
      />
    )

    expect(wrapper.state(`loadingCellClusters`)).toEqual(true)

    await wrapper.instance().componentDidMount()
    await fetchMock.flush(true) // necessary to wait for all the fetch promises to be resolved before making any assertions

    expect(wrapper.state(`cellClustersData`)).toEqual(payload)
  })

  test(`does not update its state with the JSON payload after an unsuccessful request`, async () => {
    const payload = {
      series: [`foo`, `bar`, `horse`]
    }

    fetchMock.get(`foo/json/cell-plots/` + accession + `/clusters/k/20?plotMethod=umap&` , JSON.stringify(payload))

    const wrapper = shallow(
      <TSnePlotView
        atlasUrl={`foo/`}
        metadata={metadata}
        accessKey={accessKey}
        suggesterEndpoint={`json/suggestions`}
        ks={ks}
        selectedColourBy={`20`}
        selectedColourByCategory={`clusters`}
        selectedPlotType={`umap`}
        species={species}
        experimentAccession={accession}
        selectedPlotOption={5}
        plotTypeDropdown={plotTypeDropdown}
        selectedPlotOptionLabel={``}
      />
    )

    await wrapper.instance().componentDidMount()
    await fetchMock.flush(true) // necessary to wait for all the fetch promises to be resolved before making any assertions

    expect(wrapper.state(`cellClustersData`)).not.toMatchObject(payload)
  })


  test(`the accessKey prop is used as a query parameter`, async () => {
    const payload = {
      series: [`foo`, `bar`]
    }

    fetchMock.get(`/foo/json/cell-plots/` + accession + `/clusters/k/20?plotMethod=umap&accessKey=` + accessKey + `&` , JSON.stringify(payload))

    const wrapper = shallow(
      <TSnePlotView
        atlasUrl={`foo/`}
        metadata={metadata}
        accessKey={accessKey}
        suggesterEndpoint={`json/suggestions`}
        ks={ks}
        selectedColourBy={`20`}
        selectedColourByCategory={`clusters`}
        selectedPlotType={`umap`}
        species={species}
        experimentAccession={accession}
        selectedPlotOption={5}
        plotTypeDropdown={plotTypeDropdown}
        selectedPlotOptionLabel={``}
      />
    )

    await wrapper.instance().componentDidMount()

    expect(fetchMock.called(`/foo/json/cell-plots/E-MTAB-5061/clusters/k/20?plotMethod=umap&accessKey=f60a21b8-990a-49d9-95aa-623c10865faa&`)).toEqual(true)
  })

})