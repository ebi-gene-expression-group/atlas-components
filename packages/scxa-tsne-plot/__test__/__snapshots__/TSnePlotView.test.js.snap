// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSnePlotView matches snapshot 1`] = `
<div
  className="row"
>
  <div
    className="row expanded"
    style={
      Object {
        "width": "50%",
      }
    }
  >
    <div
      className="small-12 medium-6 columns"
    >
      <label>
        Plot type:
      </label>
      <div
        className=" css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-15r8ym2-Control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-g1d714-ValueContainer"
          >
            <div
              className=" css-1uccc91-singleValue"
            >
              UMAP
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-2-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className="sc-gsnTZi ebinUc"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="small-12 medium-6 columns"
    >
      <label>
        Plot options:
      </label>
      <div
        className=" css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-15r8ym2-Control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-g1d714-ValueContainer"
          >
            <div
              className=" css-1uccc91-singleValue"
            >
              
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-3-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className="sc-gsnTZi ebinUc"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="small-12 medium-6 columns"
  >
    <div
      className="row"
    >
      <label>
        Colour plot by:
      </label>
      <div
        className=" css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-15r8ym2-Control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-g1d714-ValueContainer"
          >
            <div
              className=" css-1wa3eu0-placeholder"
            >
              Select...
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-4-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className="sc-gsnTZi ebinUc"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        className="small-12 columns"
      >
        <div />
      </div>
      <div
        className="sc-bczRLJ dMcFoN"
      >
        <p>
          Loading, please wait...
        </p>
        <img
          src="test-file-stub"
        />
      </div>
    </div>
  </div>
  <div
    className="small-12 medium-6 columns"
  >
    <div
      style={
        Object {
          "visibility": "visible",
        }
      }
    >
      <div
        className="row margin-bottom-large"
      >
        <label>
          Gene ID, gene name or gene feature:
        </label>
        <div
          style={
            Object {
              "display": "",
            }
          }
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            autoComplete="off"
            name="geneId"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="row"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        className="small-12 columns"
      >
        <div />
      </div>
      <div
        className="sc-bczRLJ dMcFoN"
      >
        <p>
          Loading, please wait...
        </p>
        <img
          src="test-file-stub"
        />
      </div>
    </div>
  </div>
</div>
`;
